### DISCUSSION TITLE
How to disable updates?

### DISCUSSION THREAD

**Discussion by Mystic8b** [ohmyzsh/ohmyzsh#10584] [Help] [answered]
Greetings, dear ones!
Oh no, of course I read the documentation. But I'm not sure if I did everything right.
I added `zstyle ':omz:update' mode disabled` to `.zshrc` before `export ZSH="/home/mystic/.oh-my-zsh"`. Did I understand the phrase "**before Oh My Zsh is loaded**" correctly?
But I have added this setting via ansible to all my machines. And after ssh login to some of them, omz is still updated automatically when opening an SSH session.
What am I doing wrong?

**Comment by mcornella** [comment, marked as answer]
That should be enough. To be clear, before Oh My Zsh is loaded means before it is `source`d (`source $ZSH/oh-my-zsh.sh`). Are you sure the line is being applied correctly? Can I see a sample zshrc file?

**Reply by Mystic8b** [reply]
> That should be enough. To be clear, before Oh My Zsh is loaded means before it is `source`d (`source $ZSH/oh-my-zsh.sh`). Are you sure the line is being applied correctly? Can I see a sample zshrc file?

Sure
[.zshrc.txt](https://github.com/ohmyzsh/ohmyzsh/files/7862321/default.zshrc.txt)

I'm not sure at what stage these settings are applied. If it is enough just to make these changes to the config, then it does not work. But if you need to run a shell to apply the changes - apparently everything is in order, as I said, I made the changes through ansible.

**Reply by mcornella** [reply]
The zshrc file looks fine to me, though I'm not sure how powerlevel10k's instant prompt interacts with the update process (at worst, it should make it *not* update, not the other way around).

You can try to run `zsh -xvic exit &> ~/omz-debug.log` to capture a trace of the start and post it here so we can see what's going on when it updates.

**Reply by Mystic8b** [reply]
Nothing happens) Is that good?
![image](https://user-images.githubusercontent.com/35394377/149348224-66eca3ac-bbe2-4d41-abd2-b445c9e42ed2.png)

**Reply by mcornella** [reply]
Yeah the trace is stored in `~/omz-debug.log`. You should upload that file or send it to <marc@ohmyz.sh>.

**Reply by Mystic8b** [reply]
Sorry, I didn't notice the redirect and just copied it. Looks like it's time for me to go home.
Sent by mail.

**Reply by mcornella** [reply]
Seems to work correctly:

```sh
# Get user's update preferences
#
# Supported update modes:
# - prompt (default): the user is asked before updating when it's time to update
# - auto: the update is performed automatically when it's time
# - reminder: a reminder is shown to the user when it's time to update
# - disabled: automatic update is turned off
zstyle -s ':omz:update' mode update_mode || update_mode=prompt
+/home/user/.oh-my-zsh/tools/check_for_upgrade.sh:13> zstyle -s :omz:update mode update_mode

# Support old-style settings
[[ "$DISABLE_UPDATE_PROMPT" != true ]] || update_mode=auto
+/home/user/.oh-my-zsh/tools/check_for_upgrade.sh:16> [[ '' != true ]]
[[ "$DISABLE_AUTO_UPDATE" != true ]] || update_mode=disabled
+/home/user/.oh-my-zsh/tools/check_for_upgrade.sh:17> [[ '' != true ]]

# Cancel update if:
# - the automatic update is disabled.
# - the current user doesn't have write permissions nor owns the $ZSH directory.
# - git is unavailable on the system.
if [[ "$update_mode" = disabled ]] \
|| [[ ! -w "$ZSH" || ! -O "$ZSH" ]] \
|| ! command -v git &>/dev/null; then
  unset update_mode
  return
fi
+/home/user/.oh-my-zsh/tools/check_for_upgrade.sh:23> [[ disabled = disabled ]]
+/home/user/.oh-my-zsh/tools/check_for_upgrade.sh:26> unset update_mode
+/home/user/.oh-my-zsh/tools/check_for_upgrade.sh:27> return
```

Is this on a machine that updates regardless of the setting?

**Reply by Mystic8b** [reply]
Yes. And it happens on all machines. I can try disabling p10k on another machine where I haven't tried disabling updates yet.

**Reply by mcornella** [reply]
On this trace the update is disabled correctly. You can try to trace only `check_for_upgrade.sh` so you can see what happens on an ssh connection where it does update. The file you should patch is at `$ZSH/tools/check_for_upgrade.sh`:
```patch
diff --git a/tools/check_for_upgrade.sh b/tools/check_for_upgrade.sh
index 293f48ed..40ab3010 100644
--- a/tools/check_for_upgrade.sh
+++ b/tools/check_for_upgrade.sh
@@ -3,6 +3,7 @@ if [[ -f ~/.zsh-update && ! -f "${ZSH_CACHE_DIR}/.zsh-update" ]]; then
   mv ~/.zsh-update "${ZSH_CACHE_DIR}/.zsh-update"
 fi
 
+set -xv
 # Get user's update preferences
 #
 # Supported update modes:
@@ -24,8 +25,10 @@ if [[ "$update_mode" = disabled ]] \
    || [[ ! -w "$ZSH" || ! -O "$ZSH" ]] \
    || ! command -v git &>/dev/null; then
   unset update_mode
+  echo cancelled && set +xv
   return
 fi
+echo continued && set +xv
 
 function current_epoch() {
   zmodload zsh/datetime
```
After that, login again and see what happens when it updates.

**Reply by Mystic8b** [reply]
Output after making changes and re-logging in via ssh.
On the other two machines, I disabled p10k in the zshrc config and the update did not happen on them. But I'm not sure that I haven't visited them for a week.
On another machine, I tried disabling p10k's instant prompt only and got the notification [oh-my-zsh] Would you like to update? [Y/n] n instead of automatic update. I don't know if we can draw any conclusions from this...

```
# Get user's update preferences
#
# Supported update modes:
# - prompt (default): the user is asked before updating when it's time to update
# - auto: the update is performed automatically when it's time
# - reminder: a reminder is shown to the user when it's time to update
# - disabled: automatic update is turned off
zstyle -s ':omz:update' mode update_mode || update_mode=prompt
+/home/user/.oh-my-zsh/tools/check_for_upgrade.sh:14> zstyle -s :omz:update mode update_mode

# Support old-style settings
[[ "$DISABLE_UPDATE_PROMPT" != true ]] || update_mode=auto
+/home/user/.oh-my-zsh/tools/check_for_upgrade.sh:17> [[ true != true ]]
+/home/user/.oh-my-zsh/tools/check_for_upgrade.sh:17> update_mode=auto
[[ "$DISABLE_AUTO_UPDATE" != true ]] || update_mode=disabled
+/home/user/.oh-my-zsh/tools/check_for_upgrade.sh:18> [[ '' != true ]]

# Cancel update if:
# - the automatic update is disabled.
# - the current user doesn't have write permissions nor owns the $ZSH directory.
# - git is unavailable on the system.
if [[ "$update_mode" = disabled ]] \
   || [[ ! -w "$ZSH" || ! -O "$ZSH" ]] \
   || ! command -v git &>/dev/null; then
  unset update_mode
  echo cancelled && set +xv
  return
fi
+/home/user/.oh-my-zsh/tools/check_for_upgrade.sh:24> [[ auto = disabled ]]
+/home/user/.oh-my-zsh/tools/check_for_upgrade.sh:25> [[ ! -w /home/user/.oh-my-zsh || ! -O /home/user/.oh-my-zsh ]]
+/home/user/.oh-my-zsh/tools/check_for_upgrade.sh:26> command -v git
echo continued && set +xv
+/home/user/.oh-my-zsh/tools/check_for_upgrade.sh:31> echo continued
continued
+/home/user/.oh-my-zsh/tools/check_for_upgrade.sh:31> set +xv
~
```

**Reply by mcornella** [reply]
It looks like you set `DISABLE_UPDATE_PROMPT`, which sets the update mode to `auto`:
```sh
# Support old-style settings
[[ "$DISABLE_UPDATE_PROMPT" != true ]] || update_mode=auto
+/home/user/.oh-my-zsh/tools/check_for_upgrade.sh:17> [[ true != true ]]
+/home/user/.oh-my-zsh/tools/check_for_upgrade.sh:17> update_mode=auto
```
That setting means *update without asking*. With the confusion you might have wanted to set `DISABLE_AUTO_UPDATE`.

This setting takes precedence over the `zstyle` one, though it being older it should be ignored if `zstyle` is set (that is a bug).

In any case, you should get rid of that setting, as it is deprecated, and just use the `zstyle` `mode disabled` setting.

**Reply by Mystic8b** [reply]
This parameter is commented out in zshrc. Or should it be removed elsewhere?

**Reply by Mystic8b** [reply]
It looks like the problem is in p10k.
```
❯ grep -rni "DISABLE_UPDATE_PROMPT" .
./.oh-my-zsh/custom/themes/powerlevel10k/internal/p10k.zsh:6336:  DISABLE_UPDATE_PROMPT=true
```

**Reply by mcornella** [reply]
Opened #10587 to track this bug.

**Reply by mcornella** [reply]
> It looks like the problem is in p10k.
> 
> ```
> ❯ grep -rni "DISABLE_UPDATE_PROMPT" .
> ./.oh-my-zsh/custom/themes/powerlevel10k/internal/p10k.zsh:6336:  DISABLE_UPDATE_PROMPT=true
> ```

That explains it, thanks!

**Reply by Mystic8b** [reply]
Thank you too! All the best.

**Comment by dphans** [comment]
In my opinion, the best UX is to just show a notification of new updates (like what homebrew is using), users who want to update can use `omz update` manually.

Showing a prompt while bash starts is annoying:
- Automation tools (like run test-cases, run debug,... in vscode) suddenly fail when encountering the prompt.
- Fast command typing operations (for me) are affected.

**Reply by mcornella** [reply]
Hey that's an option, just use `mode reminder` https://github.com/ohmyzsh/ohmyzsh?tab=readme-ov-file#getting-updates

**Reply by bkusenda** [reply]
Please do this!  The prompt is painful for automation and a potential security hole anyway.  Suppose your code gets compromised!

**Reply by mcornella** [reply]
I'm more worried about automatic security patches not getting installed than about a compromise of our repository. We have extensive system and personal security controls for it.

But if we can bypass update settings when there's a security advisory (in our roadmap to be implemented soon), we could make the default a reminder.

That said, we already have logic to bypass the update prompt if any characters where written to input before the prompt shows up, it seems it doesn't work for all cases, but it'd be good to have bug reports for that to make it better.

**Comment by buraksezer** [comment]
I think the default value of `DISABLE_UPDATE_PROMPT` should be `true` because it’s not good to interrupt users with an extension during an intense work session.

```
[oh-my-zsh] Would you like to update? [Y/n]
[oh-my-zsh] You can update manually by running `omz update`
```

This is just bad UX.
